### ISSUE TITLE
Propogate client internal hardlinks to server

### ISSUE THREAD

**Issue by phil-davis** [owncloud/client#3545]
If there are hardlinks inside the client ownCloud folder then it would be good to propagate them as hardlinks on the server.
e.g.
folder1/file1
folder2/file1
If the 2 "copies" of file1 are really hardlinked one to the other, then they only take up space once. It would be nice if that behavior propagated to the server.
I would use it on Windows NTFS hardlinks.
Should be just as useful for *nix clients that have "internal" hardlinks in their ownCloud folders.

My use case is for backup. We use a Powershell script that we have developed:
https://github.com/International-Nepal-Fellowship/Windows-Tools/tree/master/ntfs-hardlink-backup
which in turn uses a nice "ln" utility for Windows by  Hermann Schinagl:
http://schinagl.priv.at/nt/ln/ln.html
This makes incremental backups that look like full backups by using hard links to files that have not changed.
If the hard links propagated up to the ownCloud server then I could use this backup system to make backups of local data on our servers into an ownCloud folder and it would all just propagate efficiently up into the cloud, including backup history...

Anyone else got a use case for this?

**Comment by MTRichards** [comment]
Would love to hear more use cases if they are out there!

**Comment by ogoffart** [comment]
Closing as duplicate of #3545: It should be solve if we use checksum to detect copies and make a copy on the server.
Hardlinks are just a subset of copies anyway.  And copies are much more commons than hardlinks.

**Comment by phil-davis** [comment]
This is #3545 - reference points to itself. Can you post the reference to "make a copy on the server"?

**Comment by ogoffart** [comment]
I meant #3422, sorry, bad copy-paste

**Comment by phil-davis** [comment]
That #3422 is talking about minimizing up/down load by doing some extra checksums or... when a file has appeared on the client and the time stamps/size match with the server, and if the data checksums match then not to bother transferring the file in either direction server<->client. Thus allowing local "seeding" of a (new) client without the client having to actually download everything after its first connection. Certainly "a good thing".
But here I am hoping for hard-links on the server that reflect hard-links on the client. That would require either:
a) explicitly parsing the hard link structure and imitating it between client<->server, or
b) be more general and detect multiple copies of files that have the same data content in the client folders (whether because of a hardlink or just that they were copy-pasted into multiple places). Store the data content only once on the server, with multiple file entries in server folders that all point to the single copy of the data content (that is called hardlinks!). That would amount to a data de-duplication feature on the server end.

**Comment by ogoffart** [comment]
I suppose data deduplication would make sens to be implemented in the server.  Please open an issue on the core buck tracker.

**Comment by phil-davis** [comment]
I found https://github.com/owncloud/core/issues/3030
That made me think more. If the server host file system can do the dedup automagically then it fixes the wasted server storage space issue. But in that case I think the client would still upload the file, then the server back-end storage would see that the file already existed and thus not store the data content a second time.
I will think a moment and then put something in the core issues...
